jidea-74 Get model result via rrq

EmmaLRussell opened the pull request:
Adds [rrq](https://github.com/mrc-ide/rrq) to the api so that all model run requests are run in queued tasks. The `scenario/run`,  `scenario/status` and `scenario/results` endpoints are now fully implemented -  they spawn the task, report on its status and fetch its results respectively. A separate `model_run` method is the implementation of the task, and for now this is still just reading and returning the dummy data .

This implementation is largely based on that in orderly.runner. There is a `Queue` class which manages the rrq interactions, configuring the queue, and dealing with tasks. 

We need (at least) two additional docker containers to support this pattern - one for redis (which is used by rrq to manage the queue) and at least one more for the worker(s) which execute the tasks on the queue. 

Both the main api container and the worker(s) will run the same daedalus.api image, but should be spun up with different entry points: `main` for the api, `main_worker` for the workers. 

I've modified the test connection script a bit to pull out the running of the containers so that you can do that separately for manual testing.

### To test "running" a model:
1. Run the containers with `./docker/run_containers`
2. Post a request to run a model: 
``` 
curl -H "Content-Type: application/json" -d '{"modelVersion": "0.0.1", "parameters": {}}' http://localhost:8001/scenario/run
```
You should get a successful response which includes a runId.

3. Check the status of the job, replacing `[RUN_ID]` with the runId received in 2.
```
curl http://localhost:8001/scenario/status/[RUN_ID]
```
You should see a status result indicating that the job completed successfully. 

4. Get the results of the job:
```
curl http://localhost:8001/scenario/results/[RUN_ID]
```
You should see the hardcoded dummy model results returned. 

5. You can tear down the containers with `./docker/clear_containers`


### Notes
I've modified the status and results routes to take the run id as the final part of the route rather than as a query string param, as that seems a little nicer/more RESTful. 

I found that I had to downgrade the base docker image (to the version known to work for orderly.runner), as I was getting the following error: `unable to load shared object '/usr/local/lib/R/site-library/openssl/libs/openssl.so':\n  libssl.so.1.1: cannot open shared object file: No such file or directory.` - even though the shared object file did exist at the path shown. 

I couldn't crack getting mocking to work inside the Queue constructor, so the queue unit tests only mock  within the other methods.

EmmaLRussell commented on tests/testthat/helper-daedalus-api.R:
```diff
@@ -4,3 +4,23 @@ daedalus_api_endpoint <- function(..., validate = TRUE) {
   )
 }
+
+skip_if_no_redis <- function() {
+  # Allows skipping redis/rrq e2e tests on mac os
```
Pattern stolen from orderly.runner - docker isn't installed on mac os so in that case we don't attempt to run a redis container, or any of the tests which require it.

richfitz commented on .github/workflows/R-CMD-check.yaml:
```diff
@@ -45,7 +45,9 @@ jobs:
           needs: check
-
+      - name: Run redis
+        if: ${{ matrix.config.os != 'macos-latest' }}
```
We should just drop non-Linux builds from the matrix, we will never run the server on these anyway

EmmaLRussell replied:
Fine by me!

richfitz commented on .lintr:
```diff
@@ -3,16 +3,29 @@ linters: all_linters(
 	)
 exclusions: list(
-	"tests/testthat.R" = list(
+    "tests/testthat.R" = list(
```
It might be worth doing a quick tab-to-spaces translation on this file?

richfitz commented on R/api.R:
```diff
@@ -72,24 +75,27 @@ metadata <- function() {
+    data$parameters,
+    model_version = data$modelVersion
+  )
+  json_verbatim(list(runId = run_id))
```
I don't think you want json_verbatim here but a conversion to json? this would hopefully become apparent with a test :)

EmmaLRussell replied:
This is getting tested by the e2e test but i guess not detected by codecov..

The `json_verbatim` util does the auto unboxing, as I was finding that without that, the runId value got wrapped in an array. Should I be doing something else to stop that from happening? What do you mean by conversion to json? Using jsonlite or something else?

richfitz replied:
```suggestion
  jsonlite::toJSON(list(runId = jsonlite::unbox(run_id)))
```

or more simply:

```suggestion
  list(runId = jsonlite::unbox(run_id))
```

richfitz commented on R/main.R:
```diff
@@ -17,5 +17,17 @@ Options:
+
+
+main_worker <- function() {
+  queue <- Queue$new(configure_queue = FALSE)
+
+  worker <- rrq::rrq_worker$new(
+    queue$controller$queue_id,
+    con = queue$controller$con
+  )
```
```suggestion
  worker <- rrq::rrq_worker$new(
    get_queue_id(),
    con = queue$controller$con
  )
```

or is there some side-effect from constructing `Queue` you're also trying to hit?

EmmaLRussell replied:
Right, yeah I guess the worker doesn't need to make a queue - they will ned a redis connection, but should be able to make that here similarly, using `get_redis_host`

pratikunterwegs reviewed (commented):
Thanks @EmmaLRussell for adding this, looks like a fairly complex bit of work that it would have taken me a good long while if I'd tried it myself.

I've just seen that Rich has reviewed this and I don't have any contributions over that, so I've mostly looked into how this is organised so I contribute to future changes and reviews.

Following the instructions to run the containers worked for me.

One thing I didn't get immediately - I saw that the sample response JSONs under `json` have been removed, so where are the scenario results coming from now?

Just one other comment on the `.lintr` file - we're having to turn off the same linters across multiple files so I'm happy to move those to the main exclusions list for this project.

EmmaLRussell commented:
> One thing I didn't get immediately - I saw that the sample response JSONs under json have been removed, so where are the scenario results coming from now?

@pratikunterwegs The sample JSON for the run and status endpoints have been removed because we're now returning the real task run id and real task status responses. The only remaining sample json file is for the scenario results themselves, which are still being read in and returned, but now that's happening in an rrq task rather than immediately in the results endpoint.

EmmaLRussell commented:
As discussed, I've:
- removed queue configuration from the Queue constructor, and put that in a separate method in `main.R`, which, in the `./docker/run_containers` script, is called from a separate run of the api container. 

- made this script and the tests generate a new queue id each time they run.

- put a TODO in Queue to check if queue is configured, once this available in rrq

- added a test in `test-api.R` to check if `json_verbatim` is required to avoid runId coming back in an array. If you change the final line in `scenario_run` to `jsonlite::toJSON(list(runId = run_id))` rather than using `json_verbatim` (I think this what you meant) - then you get this error: `obj$runId has type 'list', not 'character'. `

richfitz commented on R/api.R:
```diff
@@ -72,24 +73,27 @@ metadata <- function() {
+#'   GET /scenario/status/<run_id:string> => json(scenarioStatus)
+#'   state queue :: queue
+scenario_status <-  function(queue, run_id) {
+  json_verbatim(queue$get_run_status(run_id))
```
```suggestion
  lapply(queue$get_run_status(run_id), jsonlite::unbox)
```

which was working before because json_verbatim was unboxing things. I'm a bit curious now about the intent of that function as I thought that was mostly concerned with splicing json...

EmmaLRussell replied:
As discussed, sounds like there's a couple of competing `json_verbatim`s - this is the one I'm using here, which I initially marked as "probably temporary" because it was getting used to re-serialise json read in from the sample files - but then proved useful for other things... It auto unboxes and sets NULLs to null. 
https://github.com/jameel-institute/daedalus.api/blob/main/R/util.R#L25

Maybe the auto-unboxing is dangerous though, as I think that will render single-value arrays as scalars, which you won't always want..

So, I could either rename this to something more sensible (`json_auto_format`??) or could take the approach you suggest of always being explicit about formatting requirements in each serialise. What do you think?

richfitz reviewed (changes requested):
these changes should end the json woes

EmmaLRussell commented on R/model_run.R:
```diff
@@ -1,4 +1,5 @@
   # Returning sample response for now
-  to_json(read_local_json("sample_scenario_results_response.json"), auto_unbox = TRUE)
+  from_file <- read_local_json("sample_scenario_results_response.json")
+  jsonlite::toJSON(from_file, auto_unbox = TRUE)
```
For reasons which I cannot explain, after the util rename, `model_run` wasn't able to find it anymore in the e2e test.... I've no idea why! But since this is going to be replaced with a call to the daedalus package soon, maybe we can just not use the util here (for now).

EmmaLRussell commented:
I tried `json_validate::json_serialise` - it worked fine for the simple cases, but was getting an error in V8 when attempting to serialise the metadata, which is the most complicated schema, (something like 'Could not convert null or undefined to object' - but there were no nulls in the object structure). 

So I've gone with the explicit unboxing, keeping the use of the renamed util for the metadata and results endpoints.

pratikunterwegs commented:
I don't have anything to add beyond Rich's review so feel free to merge anytime!
